### PR TITLE
Add SmolLM2

### PR DESCRIPTION
**Pull request by ysjprojects**
https://huggingface.co/collections/HuggingFaceTB/smollm2-6723884218bcda64b34d7db9

SOTA small language models (135M, 360M, 1.7B) by Hugging Face.

Great for on-device uses

Adding because of request on mobile-oriented models  #1846

### CODE REVIEW

**Comment by Andrei-Aksionov** [comment]
Hello @ysjprojects 

Do you think it's ready for a review or still in WIP?

**Comment by ysjprojects** [comment]
> Hello @ysjprojects
> 
> Do you think it's ready for a review or still in WIP?

just pushed some final changes, should be ready for review

**Comment by Andrei-Aksionov** [comment]
Hello @ysjprojects 
I cannot download the model.
For some reason, huggingface_hub returns more files than are shown on the model page.
One of the files is `training_args.bin`, which breaks the download code.
In other words, the code finds `.bin` files and "thinks" that's this is a model, so doesn't make an attempt to download`.safetensors`.

**Comment by ysjprojects** [comment]
> Hello @ysjprojects I cannot download the model. For some reason, huggingface_hub returns more files than are shown on the model page. One of the files is `training_args.bin`, which breaks the download code. In other words, the code finds `.bin` files and "thinks" that's this is a model, so doesn't make an attempt to download`.safetensors`.
> 
> (made a small change to download only model weights if there is .bin files.)

interesting, does the small change fix the issue or would I need to look into it?

File: litgpt/prompts.py [hunk 1]
```diff
@@ -321,6 +327,7 @@ def apply(self, prompt: str, **kwargs: str) -> str:
     "olmo": OLMo,
     "qwen2.5": Qwen2_5,
     "qwq": QwQ,
+    "smollm2": SmolLM2, # SmolLM uses a different template
```
Comment (Andrei-Aksionov) [inline litgpt/prompts.py:330]: Do we really need this comment here?

File: tests/test_model.py [hunk 2]
```diff
@@ -910,6 +911,66 @@ def test_against_original_salamandra(model_name, device, dtype):
     ours_y = ours_model(x)
     theirs_y = theirs_model(x)["logits"].to(dtype)  # HF converts logits to float
     torch.testing.assert_close(ours_y, theirs_y)
+    
+
+@torch.inference_mode()
+@pytest.mark.parametrize("model_name", ("SmolLM2-135M", "SmolLM2-1.7B"))
```
Comment (Andrei-Aksionov) [inline tests/test_model.py:917]: Add 360M to the mix

**Comment by Andrei-Aksionov** [comment]
Hello @ysjprojects 

> interesting, does the small change fix the issue or would I need to look into it?

Yes, the fix did work. 

The PR looks great. Thanks 😊 

There are only a couple of nits.

**Comment by ysjprojects** [comment]
> Hello @ysjprojects
> 
> > interesting, does the small change fix the issue or would I need to look into it?
> 
> Yes, the fix did work.
> 
> The PR looks great. Thanks 😊
> 
> There are only a couple of nits.

cool! have resolved the requested changes

**Pull request merged**